refactor(migrations): migrate toArray and get methods in query migration

devversion opened the pull request:
See individual commits

devversion committed:
refactor(migrations): provide AST path for resolved template and host references

This allows us to detect usages in templates and figure out if they are
e.g. invoking a problematic method like `.pipe` for the output
migration, or in queries `.toArray` etc.

devversion committed:
refactor(migrations): migrate `toArray` and `get` methods in query migration

This commit migrates `toArray` and `get` methods of `QueryList`. The
`toArray` method is no longer needed for multi-queries, and `get` can be
replaced with native array `at`.

JoostK commented on packages/core/schematics/migrations/signal-migration/src/passes/reference_resolution/reference_kinds.ts:
```diff
@@ -40,6 +40,8 @@ export interface TemplateReference<D extends ClassFieldDescriptor> {
     node: TmplAstNode;
     /** Expression AST node that represents the reference. */
     read: PropertyRead;
+    /** Expression AST sequentially visited to reach the given read. */
```
Might be useful to mention this follows top-down ordering, not bottom-up, making `read` the _last_ element in the array.

May also be useful to mention how this is a mitigation for the lack of parent pointers for `AST` nodes, as having those would make this path unnecessary.

devversion replied:
can we have parent pointers please 🙏 😄  cc. @crisbeto

crisbeto replied:
I don't see why not 🤷 we don't have them right now because we haven't needed them.

devversion replied:
I will make a note. Shouldn't be too difficult to add. Only problem I see is serialization; but not sure if anything serializes the AST.

devversion replied:
Interesting. I believe they could use a JSON stringify transform to omit the "parent" pointes.

JoostK commented on packages/core/schematics/migrations/signal-queries-migration/fn_get_replacement.ts:
```diff
@@ -0,0 +1,76 @@
+    return;
+  }
+
+  if (!globalMetadata.knownQueryFields[ref.target.key].isMulti) {
```
Is there a possibility of this indexed access evaluating to `undefined`? TS won't catch this statically so such failure mode may have gone unobserved?

devversion replied:
there isn't, but we can add a safety check. Good point

JoostK reviewed (commented):
🚗

devversion committed:
fixup! refactor(migrations): provide AST path for resolved template and host references

Address feedback

devversion committed:
fixup! refactor(migrations): migrate `toArray` and `get` methods in query migration

Address feedback

devversion committed:
refactor(migrations): replace remaining `performance.now()` usages.

Those usages break in the Angular CLI schematic execution, so we should
replace them as well.

devversion commented:
This PR was merged into the repository by commit 2c8449a2f5f21a4b40f542141ff03362f6a4e71a.

The changes were merged into the following branches: main
